scipy/sparse/_coo.py

```diff
@@ -556,31 +571,94 @@ def _add_dense(self, other):
+        if self.ndim < 3:
+            M, N = self._shape_as_2d
+            coo_todense(M, N, self.nnz, self.row, self.col, self.data,
+                        result.ravel('A'), fortran)
+        else:
+            coords = np.concatenate(self.coords)
+            coo_todense_nd(self.shape, self.nnz, self.ndim, coords,
+                           self.data, result.ravel('A'), fortran)
         return self._container(result, copy=False)
+    
 
-    def _matmul_vector(self, other):
-        result_shape = self.shape[0] if self.ndim > 1 else 1
-        result = np.zeros(result_shape,
-                          dtype=upcast_char(self.dtype.char, other.dtype.char))
+    def _add_sparse(self, other):
+        if self.ndim < 3:
+            return _data_matrix._add_sparse(self, other)
 
-        if self.ndim == 2:
-            col = self.col
-            row = self.row
-        elif self.ndim == 1:
-            col = self.coords[0]
-            row = np.zeros_like(col)
-        else:
-            raise NotImplementedError(
-                f"coo_matvec not implemented for ndim={self.ndim}")
+        if other.shape != self.shape:
+            raise ValueError(f'Incompatible shapes ({self.shape} and {other.shape})')
+        other = self.__class__(other)
+        new_data = np.concatenate((self.data, other.data))
+        new_coords = tuple(np.concatenate((self.coords, other.coords), axis=1))
+        A = self.__class__((new_data, new_coords), shape=self.shape)         
+        return A
+
+    
+    def _sub_sparse(self, other):
+        if self.ndim < 3:
+            return _data_matrix._sub_sparse(self, other)
+
+        if other.shape != self.shape:
+            raise ValueError(f'Incompatible shapes ({self.shape} and {other.shape})')
+        other = self.__class__(other)
+        new_data = np.concatenate((self.data, -other.data))
+        new_coords = tuple(np.concatenate((self.coords, other.coords), axis=1))
+        A = coo_array((new_data, new_coords), shape=self.shape)
+        return A
+    
+
+    def _matmul_vector(self, other):
+        if self.ndim < 3:
+            result_shape = self.shape[0] if self.ndim > 1 else 1
+            result = np.zeros(result_shape,
+                              dtype=upcast_char(self.dtype.char, other.dtype.char))
+            if self.ndim == 2:
+                col = self.col
+                row = self.row
+            elif self.ndim == 1:
+                col = self.coords[0]
+                row = np.zeros_like(col)
+            else:
+                raise NotImplementedError(
+                    f"coo_matvec not implemented for ndim={self.ndim}")
+
+            coo_matvec(self.nnz, row, col, self.data, other, result)
+            # Array semantics return a scalar here, not a single-element array.
+            if isinstance(self, sparray) and result_shape == 1:
+                return result[0]
+            return result
+        
+        else: # if dim >= 3
```

Comment:
Let's move this condition to the top, and then because we unconditionally return, we won't need to indent the existing code (to make the diff a bit cleaner).

Reply:
Addressed at https://github.com/scipy/scipy/pull/21197/commits/46e18b70301c944d56cf8c68aac6209be6f73291.